### PR TITLE
gh-109945: Enable spec of multiple curves/groups for TLS

### PR DESCRIPTION
**Pull request by planetf1**
* Issue #109945

Co-authored-by: Martin Schmatz

This change makes it possible to allow multiple groups to be specified in a colon separated string. See issues for more detail

The implementation choice is to modify the existing function rather than introduce a new one.

**Changes:**

 - add new logic to set SSL curves when OpenSSL >=3 
 - add additional testcases to verify argument format (check exception)
 - add additional testcases to verify multiple curves if OpenSSL >=3 
 - added Blurb
 - Added docs
 
**Current Status:**

- Ready for review

Addressed:
- [x] Initially opening as draft to check contributor guidelines, code layout etc.
- [x] Some further testcases are commented out which check format. These caused an exception I need to check
- [x] some sanitizer tests failing - to be investigated

**Note to reviewers:**

 - in additional to normal checks please see closely the explanation of how the new function responds to bad curves in the OpenSSL>3 cases - is this ok?
 - I noticed additional files created in `Lib/lib2to3/` which I presume should not be checked in
 - I can squash changes to one/less commit as/if required (or maintainer can squash when merged)
 

* Issue: gh-109945

### CODE REVIEW

**Comment by planetf1** [comment]
FYI am aware of the test failure. Will be working on update for that. (apologies for delay) + arranging CLA.

**Comment by planetf1** [comment]
The test failure occurs within *test_set_ecdh_curve* which tests a follows:
```
    @unittest.skipUnless(ssl.HAS_ECDH, "ECDH disabled on this OpenSSL build")
    def test_set_ecdh_curve(self):
        ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
        ctx.set_ecdh_curve("prime256v1")
        ctx.set_ecdh_curve(b"prime256v1")
        self.assertRaises(TypeError, ctx.set_ecdh_curve)
        self.assertRaises(TypeError, ctx.set_ecdh_curve, None)
        self.assertRaises(ValueError, ctx.set_ecdh_curve, "foo")
        self.assertRaises(ValueError, ctx.set_ecdh_curve, b"foo")
```

This raises two issues

a) If the change in this PR is to be included, the unit tests should be updated to also include lists of curves - no problem

More relevant is

b) The current tests assert that an exception is raised if an invalid curve is specified (which is not in the initial code)

There are a few ways of handling this

1. I could parse the list of curves, checking each one for validity and raising an exception either on first failure, or after finding all errors. (ie strtok loop, OBJ_snnid checks). This would keep the behaviour effectively the same as the current version, with precisely reported errors, but is the most complex. (and some concern about any loop iterating through the list ad implementation questions, such as is using of string.h / strtok_r ok?)
2. There are lots of places in the ssl code in general where we use _setSSLError and just return NULL without formatting a python error - even in the original code this occurs. The test could be changed to permit this -- but this is a change in behaviour
3. In addition to _setSSLError when we fail to set the curves, I could add something like `PyErr_Format(PyExc_ValueError,"unknown elliptic curves %R", name);`. This keeps the behaviour closest for the test case, and is the simplest implementation, but does mean other ssl failures now raise an exception where they didn't previously.

Not being very familiar with the code I'm interested in feedback as to what the most appropriate way of handling this would be?

**Comment by planetf1** [comment]
The test failure occurs within *test_set_ecdh_curve* which tests a follows:
```
    @unittest.skipUnless(ssl.HAS_ECDH, "ECDH disabled on this OpenSSL build")
    def test_set_ecdh_curve(self):
        ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
        ctx.set_ecdh_curve("prime256v1")
        ctx.set_ecdh_curve(b"prime256v1")
        self.assertRaises(TypeError, ctx.set_ecdh_curve)
        self.assertRaises(TypeError, ctx.set_ecdh_curve, None)
        self.assertRaises(ValueError, ctx.set_ecdh_curve, "foo")
        self.assertRaises(ValueError, ctx.set_ecdh_curve, b"foo")
```

This raises two issues

a) If the change in this PR is to be included, the unit tests should be updated to also include lists of curves - no problem

More relevant is

b) The current tests assert that an exception is raised if an invalid curve is specified (which is not in the initial code)

There are a few ways of handling this

1. I could parse the list of curves, checking each one for validity and raising an exception either on first failure, or after finding all errors. (ie strtok loop, OBJ_snnid checks). This would keep the behaviour effectively the same as the current version, with precisely reported errors, but is the most complex. (and some concern about any loop iterating through the list ad implementation questions, such as is using of string.h / strtok_r ok?)
2. There are lots of places in the ssl code in general where we use _setSSLError and just return NULL without formatting a python error - even in the original code this occurs. The test could be changed to permit this -- but this is a change in behaviour
3. In addition to _setSSLError I could add something like `PyErr_Format(PyExc_ValueError,"unknown elliptic curves %R", name);`. This keeps the behaviour closest for the test case, but does mean other ssl failures now raise an exception where they didn't previously.

Not being very familiar with the code I'm interested in feedback as to what the most appropriate way of handling this would be?

**Comment by planetf1** [comment]
The current PR proposal builds cleanly. However is an area that needs discussion

Previously an invalid (no NID) curve would result in a value exception, whilst any other kind of SSL error would not, though it would set the ssl error. This passed tests cleanly.

Since we now can have multiple curves, one option is to simple call the underlying SSL function to set the curves. This makes it difficult to distinguish between the two cases, meaning that the code returns a valueerror in both cases (for invalid curves)

In working through this, I also noticed that the thread sanitizer checks FAIL when _setSSLError() is called. Before this code change we still did call this, just in less cases, and via a code path that the unix tests didn't check. With the code change the new code (minus the most recent commit) would set this on every exception (as above). My conclusion is that this function isn't thread safe? So is that a more general bug that needs investigation?

**Comment by planetf1** [comment]
Rebased. Would very much appreciate any review comments . Thanks!

**Comment by planetf1** [comment]
Aware of one doc validation error from my news blurb

*Adds support for multiple curves to be specified in SSLContext.set_ecdh_curve() for OpenSSL 3.0 and above by setting curve_name to a colon separated list of curves. This allows multiple curves to be passed on a TLS client hello.*

/home/runner/work/cpython/cpython/build/NEWS:70: WARNING: py:func reference target not found: warnings.filterswarnings

File: Lib/test/test_ssl.py [hunk 1]
```diff
@@ -1375,11 +1375,29 @@ def test_set_ecdh_curve(self):
...
         self.assertRaises(ValueError, ctx.set_ecdh_curve, b"foo")
-
+        # Multiple bad curves should cause error for any OpenSSL version
+        self.assertRaises(ValueError, ctx.set_ecdh_curve, "foo:bar")
+        self.assertRaises(ValueError, ctx.set_ecdh_curve, b"foo:bar")
+        self.assertRaises(ValueError, ctx.set_ecdh_curve, "prime256v1:bar")
+        self.assertRaises(ValueError, ctx.set_ecdh_curve, b"prime256v1:bar")
+        self.assertRaises(ValueError, ctx.set_ecdh_curve, "foo:prime256v1")
+        self.assertRaises(ValueError, ctx.set_ecdh_curve, b"foo:prime256v1")
+        #self.assertRaises(ValueError, ctx.set_ecdh_curve, ":")
```
Comment (gpshead) [inline Lib/test/test_ssl.py:1393]: what's up with the commented out test cases?  are these not valuable?  it seems like a set of edge cases worth covering to define behavior on unusual inputs.

**Comment by gpshead** [comment]
FYI - I'm looping in @sethmlarson for additional eyeballs on whether or not they think this makes sense given it is a `_ssl` module change.

**Comment by sethmlarson** [comment]
Going to also ping @woodruffw and @jvdprng since they're working on an adjacent project.

File: Modules/_ssl.c [hunk 2]
```diff
@@ -4399,8 +4400,10 @@ _ssl__SSLContext_set_ecdh_curve(PySSLContext *self, PyObject *name)
     SSL_CTX_set_tmp_ecdh(self->ctx, key);
     EC_KEY_free(key);
 #else
-    if (!SSL_CTX_set1_groups(self->ctx, &nid, 1)) {
-        _setSSLError(get_state_ctx(self), NULL, 0, __FILE__, __LINE__);
+    int res = SSL_CTX_set1_groups_list(self->ctx, PyBytes_AS_STRING(name_bytes));
```
Comment (sethmlarson) [inline Modules/_ssl.c:4403]: SSL_CTX_set1_groups_list also supports another syntax, adding a `?` before the curve name makes it "optional", [quoting the docs](https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set1_groups_list.html):

> If a group name is preceded with the ? character, it will be ignored if an implementation is missing.

This isn't tested in our test suite, it would be useful to test that so future contributors know that syntax exists.

File: Lib/test/test_ssl.py [hunk 3]
```diff
@@ -1375,11 +1375,29 @@ def test_set_ecdh_curve(self):
         ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
         ctx.set_ecdh_curve("prime256v1")
         ctx.set_ecdh_curve(b"prime256v1")
+        # Only OpenSSL 3 and above supported for multiple curves
+        if (IS_OPENSSL_3_0_0 >= 3):
```
Comment (sethmlarson) [inline Lib/test/test_ssl.py:1379]: `IS_OPENSSL_3_0_0` is a boolean, meaning this branch will always fail?

File: Doc/library/ssl.rst [hunk 4]
```diff
@@ -1769,6 +1769,10 @@ to speed up repeated connections from the same clients.
    a well-known elliptic curve, for example ``prime256v1`` for a widely
    supported curve.
 
+   For OpenSSL 3.0 and above *curve_name* parameter can be a colon separated
```
Comment (sethmlarson) [inline Doc/library/ssl.rst:1772]: This will need a "new in version 3.14" admonition.

**Review by sethmlarson — CHANGES_REQUESTED** [review]

**Comment by woodruffw** [comment]
> Going to also ping @woodruffw and @jvdprng since they're working on an adjacent project.

Thank you for the ping!

I'd like to understand the motivation a little better here: is the current cipher suite configuration insufficient? And is there a reason why `ssl` needs an explicit configuration for this, instead of allowing the client/server to negotiate a compatible cipher suite?

**Comment by sethmlarson** [comment]
> I'd like to understand the motivation a little better here: is the current cipher suite configuration insufficient? And is there a reason why `ssl` needs an explicit configuration for this, instead of allowing the client/server to negotiate a compatible cipher suite?

This API is for configuring the curves to offer during ECDH key exchange, not the ciphers.

**Comment by woodruffw** [comment]
> 
> This API is for configuring the curves to offer during ECDH key exchange, not the ciphers.

Whoops!

That makes more sense, although I'm still curious about the intended application here -- IME it's not common for user-level TLS APIs (with the notable exception of OpenSSL) to expose curve selection as a configurable parameter, instead preferring to keep that as a implementation detail of negotiation (with the assumption that the TLS implementation either ensures a baseline level of security *or* the user configures a higher level "security margin" setting that influences the curve selection.)

TL;DR: My basic concern is that this exposes an API that's  uncommon among TLS APIs, is a low level option that is typically only influenced by higher level APIs, and is a potential footgun. But one could easily argue that many of OpenSSL's APIs are footguns and thus there is ample precedent for exposing this 🙂

**Comment by planetf1** [comment]
Quick update - thanks for the review comments - apologies for the delay: just back from a few week's vacation, so will work through & respond specifically. Appreciated.

**Comment by ronf** [comment]
Regarding use cases, one reason for this is to be able to control whether or not a given connection will attempt to use post-quantum crypto algorithms during the key exchange (such as the hybrid X25519MLKEM768), or plain elliptic curve (such as X25519 by itself). OpenSSL 3.5 defaults to both of these, with a preference for the hybrid PQC algorithm. If an application doesn't want to pay the extra cost for PQC, they could use this mechanism to select only plain EC curves, removing the PQC algorithms from the list.

That said, we might want to think about naming the SSLContext method set_groups() rather than set_ecdh_curve() if we're going to use the OpenSSL SSL_CTX_set1_groups() function here. Alternately, we keep set_ecdh_curve() as only accepting EC curves (or maybe even jus the single curve it accepts now) and make support for multiple groups (including the fancier syntax with things like question mark and star for setting which key shares to generate) only in the new set_groups() method.

**Comment by davidkel** [comment]
@gpshead @sethmlarson @woodruffw I would say I am keen to see this in Python and I think I agree with @ronf that I think it might be better to have this as a different api rather than overloading the `set_ecdh_curve()` api as it really doesn't describe what is actually happening here. Admittedly it is governed by what OpenSSL expects and can parse and in OpenSSL 3.5 the group string that it can process has become far more powerful. As well as prefixes such as `?` and `-` there is now a prefix `*` to control which keyshares should be sent, also you can now group curves together by using a tuple separator (to group similar strength mechanisms). So it is really powerful now and allows it to be able to meet NIST requirements for using Quantumsafe Key exchange mechanisms. If you look at various network servers such as HAProxy or NGINX these provide ways to provide your own group string but just passes this to OpenSSL to process and I think People who write Python network servers will want to be able to offer the same ability through their configuration rather than changing the OpenSSL config file or trying to use an ENV var to point to another OpenSSL config file (which may require specialist permissions). 

There is a further extension to this which I think is also important and that is observability. It is becoming more important to be able to determine which group was negotiated,  which cipher was selected etc to be able to determine the level of security being used to make sure it is adequate and quantumsafe and I would try to include some more apis to access this information so a Python server/client can log this information.

I worked with @planetf1 and know he won't be returning to this, but I would like to see if I can pick it up (if anyone has some pointers around adding and testing new apis the the ssl module that would be great as I see there is some form of code generation going on) and want to guage what everyone thinks so any feedback would be great.

**Comment by ronf** [comment]
> There is a further extension to this which I think is also important and that is observability. It is becoming more important to be able to determine which group was negotiated, which cipher was selected etc to be able to determine the level of security being used to make sure it is adequate and quantumsafe and I would try to include some more apis to access this information so a Python server/client can log this information.

Agreed on this! There's an OpenSSL API SSL_get0_group_name() which can be called to get back the name of the group which was selected for performing the key agreement. Ideally, in addition to changing this code to call SSL_CTX_set1_groups(), there would be a new method which can call SSL_get0_group_name(). Note that this would be called on an SSL object, not an SSL_CTX, though. As such, it would probably need to be available as something like SSLSocket.group(), similar to the SSLSocket.cipher() method which exists today.

**Comment by ronf** [comment]
I did a little more looking into this, and came up with the following change to add a set_groups() method on SSLContext:

```patch
diff --git a/Modules/_ssl.c b/Modules/_ssl.c
index 014e624f6c2..da1ed457e82 100644
--- a/Modules/_ssl.c
+++ b/Modules/_ssl.c
@@ -3403,6 +3403,25 @@ _ssl__SSLContext_get_ciphers_impl(PySSLContext *self)
 }
 
 
+/*[clinic input]
+@critical_section
+_ssl._SSLContext.set_groups
+    grouplist: str
+    /
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLContext_set_groups_impl(PySSLContext *self, const char *grouplist)
+/*[clinic end generated code: output=0b5d05dfd371ffd0 input=2cc64cef21930741]*/
+{
+    if (!SSL_CTX_set1_groups_list(self->ctx, grouplist)) {
+        _setSSLError(get_state_ctx(self), "unrecognized group", 0, __FILE__, __LINE__);
+        return NULL;
+    }
+    Py_RETURN_NONE;
+}
+
+
 static int
 do_protocol_selection(int alpn, unsigned char **out, unsigned char *outlen,
                       const unsigned char *server_protocols, unsigned int server_protocols_len,
@@ -5249,6 +5268,7 @@ static struct PyMethodDef context_methods[] = {
     _SSL__SSLCONTEXT__WRAP_SOCKET_METHODDEF
     _SSL__SSLCONTEXT__WRAP_BIO_METHODDEF
     _SSL__SSLCONTEXT_SET_CIPHERS_METHODDEF
+    _SSL__SSLCONTEXT_SET_GROUPS_METHODDEF
     _SSL__SSLCONTEXT__SET_ALPN_PROTOCOLS_METHODDEF
     _SSL__SSLCONTEXT_LOAD_CERT_CHAIN_METHODDEF
     _SSL__SSLCONTEXT_LOAD_DH_PARAMS_METHODDEF
```

You also need to run the following to get clinic to auto-generate the necessary changes in Modules/clinic/_ssl.c.h:

```
Tools/clinic/clinic.py Modules/_ssl.c
```

More documentation on clinic can be found at https://docs.python.org/3.10/howto/clinic.html.

With this, I was able to pass a string into set_groups() which supported all of the OpenSSL delimiters/flags ('?' for optional, '*' for specifying what to add as key shares, '/' for specifying multiple group tuples at different priorities, and ':' as the separator within each group tuple). See documentation on this at https://docs.openssl.org/master/man3/SSL_CTX_set1_curves/.

This doesn't implement the "SSLSocket.group()" method discussed above, but I'm going to look into that next.

**Comment by ronf** [comment]
Here's a version which adds SSLSocket.group() to return the selected key agreement group name after the TLS handshake has completed:

```patch
diff --git a/Lib/ssl.py b/Lib/ssl.py
index 7e3c4cbd6bb..452118822c1 100644
--- a/Lib/ssl.py
+++ b/Lib/ssl.py
@@ -931,6 +931,10 @@ def cipher(self):
         ssl_version, secret_bits)``."""
         return self._sslobj.cipher()
 
+    def group(self):
+        """Return the currently selected key agreement group name."""
+        return self._sslobj.group()
+
     def shared_ciphers(self):
         """Return a list of ciphers shared by the client during the handshake or
         None if this is not a valid server connection.
@@ -1206,6 +1210,14 @@ def cipher(self):
         else:
             return self._sslobj.cipher()
 
+    @_sslcopydoc
+    def group(self):
+        self._checkClosed()
+        if self._sslobj is None:
+            return None
+        else:
+            return self._sslobj.group()
+
     @_sslcopydoc
     def shared_ciphers(self):
         self._checkClosed()
diff --git a/Modules/_ssl.c b/Modules/_ssl.c
index 014e624f6c2..bc96b7d15ca 100644
--- a/Modules/_ssl.c
+++ b/Modules/_ssl.c
@@ -2142,6 +2142,25 @@ _ssl__SSLSocket_cipher_impl(PySSLSocket *self)
     return cipher_to_tuple(current);
 }
 
+/*[clinic input]
+@critical_section
+_ssl._SSLSocket.group
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLSocket_group_impl(PySSLSocket *self)
+/*[clinic end generated code: output=9c168ee877017b95 input=5f187d8bf0d433b7]*/
+{
+    const char *group_name;
+
+    if (self->ssl == NULL)
+        Py_RETURN_NONE;
+    group_name = SSL_get0_group_name(self->ssl);
+    if (group_name == NULL)
+        Py_RETURN_NONE;
+    return PyUnicode_DecodeFSDefault(group_name);
+}
+
 /*[clinic input]
 @critical_section
 _ssl._SSLSocket.version
@@ -3023,6 +3042,7 @@ static PyMethodDef PySSLMethods[] = {
     _SSL__SSLSOCKET_GETPEERCERT_METHODDEF
     _SSL__SSLSOCKET_GET_CHANNEL_BINDING_METHODDEF
     _SSL__SSLSOCKET_CIPHER_METHODDEF
+    _SSL__SSLSOCKET_GROUP_METHODDEF
     _SSL__SSLSOCKET_SHARED_CIPHERS_METHODDEF
     _SSL__SSLSOCKET_VERSION_METHODDEF
     _SSL__SSLSOCKET_SELECTED_ALPN_PROTOCOL_METHODDEF
@@ -3402,6 +3422,23 @@ _ssl__SSLContext_get_ciphers_impl(PySSLContext *self)
 
 }
 
+/*[clinic input]
+@critical_section
+_ssl._SSLContext.set_groups
+    grouplist: str
+    /
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLContext_set_groups_impl(PySSLContext *self, const char *grouplist)
+/*[clinic end generated code: output=0b5d05dfd371ffd0 input=2cc64cef21930741]*/
+{
+    if (!SSL_CTX_set1_groups_list(self->ctx, grouplist)) {
+        _setSSLError(get_state_ctx(self), "unrecognized group", 0, __FILE__, __LINE__);
+        return NULL;
+    }
+    Py_RETURN_NONE;
+}
 
 static int
 do_protocol_selection(int alpn, unsigned char **out, unsigned char *outlen,
@@ -5249,6 +5286,7 @@ static struct PyMethodDef context_methods[] = {
     _SSL__SSLCONTEXT__WRAP_SOCKET_METHODDEF
     _SSL__SSLCONTEXT__WRAP_BIO_METHODDEF
     _SSL__SSLCONTEXT_SET_CIPHERS_METHODDEF
+    _SSL__SSLCONTEXT_SET_GROUPS_METHODDEF
     _SSL__SSLCONTEXT__SET_ALPN_PROTOCOLS_METHODDEF
     _SSL__SSLCONTEXT_LOAD_CERT_CHAIN_METHODDEF
     _SSL__SSLCONTEXT_LOAD_DH_PARAMS_METHODDEF
```

**Comment by ronf** [comment]
One more addition: The following now includes an `SSLContext.get_groups(all: bool = False)` method, paralleling get_ciphers() there. It uses SSL_CTX_get0_implemented_groups() to return a list of available compatible groups depending on the SSLContext's current minimum and maximum TLS version. The "all" argument defaults to `False`, meaning to return only the official IANA names. However, if you want to also get back a list of the older EC curve aliases, you can pass in a `True` value here, causing it to add things like "P-521" to the returned list.

```patch
diff --git a/Lib/ssl.py b/Lib/ssl.py
index 7e3c4cbd6bb..452118822c1 100644
--- a/Lib/ssl.py
+++ b/Lib/ssl.py
@@ -931,6 +931,10 @@ def cipher(self):
         ssl_version, secret_bits)``."""
         return self._sslobj.cipher()
 
+    def group(self):
+        """Return the currently selected key agreement group name."""
+        return self._sslobj.group()
+
     def shared_ciphers(self):
         """Return a list of ciphers shared by the client during the handshake or
         None if this is not a valid server connection.
@@ -1206,6 +1210,14 @@ def cipher(self):
         else:
             return self._sslobj.cipher()
 
+    @_sslcopydoc
+    def group(self):
+        self._checkClosed()
+        if self._sslobj is None:
+            return None
+        else:
+            return self._sslobj.group()
+
     @_sslcopydoc
     def shared_ciphers(self):
         self._checkClosed()
diff --git a/Modules/_ssl.c b/Modules/_ssl.c
index 014e624f6c2..13c1338d132 100644
--- a/Modules/_ssl.c
+++ b/Modules/_ssl.c
@@ -2142,6 +2142,25 @@ _ssl__SSLSocket_cipher_impl(PySSLSocket *self)
     return cipher_to_tuple(current);
 }
 
+/*[clinic input]
+@critical_section
+_ssl._SSLSocket.group
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLSocket_group_impl(PySSLSocket *self)
+/*[clinic end generated code: output=9c168ee877017b95 input=5f187d8bf0d433b7]*/
+{
+    const char *group_name;
+
+    if (self->ssl == NULL)
+        Py_RETURN_NONE;
+    group_name = SSL_get0_group_name(self->ssl);
+    if (group_name == NULL)
+        Py_RETURN_NONE;
+    return PyUnicode_DecodeFSDefault(group_name);
+}
+
 /*[clinic input]
 @critical_section
 _ssl._SSLSocket.version
@@ -3023,6 +3042,7 @@ static PyMethodDef PySSLMethods[] = {
     _SSL__SSLSOCKET_GETPEERCERT_METHODDEF
     _SSL__SSLSOCKET_GET_CHANNEL_BINDING_METHODDEF
     _SSL__SSLSOCKET_CIPHER_METHODDEF
+    _SSL__SSLSOCKET_GROUP_METHODDEF
     _SSL__SSLSOCKET_SHARED_CIPHERS_METHODDEF
     _SSL__SSLSOCKET_VERSION_METHODDEF
     _SSL__SSLSOCKET_SELECTED_ALPN_PROTOCOL_METHODDEF
@@ -3402,6 +3422,67 @@ _ssl__SSLContext_get_ciphers_impl(PySSLContext *self)
 
 }
 
+/*[clinic input]
+@critical_section
+_ssl._SSLContext.set_groups
+    grouplist: str
+    /
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLContext_set_groups_impl(PySSLContext *self, const char *grouplist)
+/*[clinic end generated code: output=0b5d05dfd371ffd0 input=2cc64cef21930741]*/
+{
+    if (!SSL_CTX_set1_groups_list(self->ctx, grouplist)) {
+        _setSSLError(get_state_ctx(self), "unrecognized group", 0, __FILE__, __LINE__);
+        return NULL;
+    }
+    Py_RETURN_NONE;
+}
+
+/*[clinic input]
+@critical_section
+_ssl._SSLContext.get_groups
+    all: bool = False
+    /
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLContext_get_groups_impl(PySSLContext *self, int all)
+/*[clinic end generated code: output=1c4eae66f237d494 input=df1e5272483615ef]*/
+{
+    STACK_OF(OPENSSL_CSTRING) *groups;
+    const char *group;
+    size_t i, num;
+    PyObject *result = NULL;
+
+    if ((groups = sk_OPENSSL_CSTRING_new_null()) == NULL) {
+        _setSSLError(get_state_ctx(self), "Can't allocate stack", 0, __FILE__, __LINE__);
+        return NULL;
+    }
+
+    if (!SSL_CTX_get0_implemented_groups(self->ctx, all, groups)) {
+        _setSSLError(get_state_ctx(self), "Can't get groups", 0, __FILE__, __LINE__);
+        sk_OPENSSL_CSTRING_free(groups);
+        return NULL;
+    }
+
+    num = sk_OPENSSL_CSTRING_num(groups);
+    result = PyList_New(num);
+    if (result == NULL) {
+        _setSSLError(get_state_ctx(self), "Can't allocate list", 0, __FILE__, __LINE__);
+        sk_OPENSSL_CSTRING_free(groups);
+        return NULL;
+    }
+
+    for (i = 0; i < num; ++i) {
+        group = sk_OPENSSL_CSTRING_value(groups, i);
+        PyList_SET_ITEM(result, i, PyUnicode_DecodeFSDefault(group));
+    }
+
+    sk_OPENSSL_CSTRING_free(groups);
+    return result;
+}
 
 static int
 do_protocol_selection(int alpn, unsigned char **out, unsigned char *outlen,
@@ -5249,6 +5330,7 @@ static struct PyMethodDef context_methods[] = {
     _SSL__SSLCONTEXT__WRAP_SOCKET_METHODDEF
     _SSL__SSLCONTEXT__WRAP_BIO_METHODDEF
     _SSL__SSLCONTEXT_SET_CIPHERS_METHODDEF
+    _SSL__SSLCONTEXT_SET_GROUPS_METHODDEF
     _SSL__SSLCONTEXT__SET_ALPN_PROTOCOLS_METHODDEF
     _SSL__SSLCONTEXT_LOAD_CERT_CHAIN_METHODDEF
     _SSL__SSLCONTEXT_LOAD_DH_PARAMS_METHODDEF
@@ -5259,6 +5341,7 @@ static struct PyMethodDef context_methods[] = {
     _SSL__SSLCONTEXT_CERT_STORE_STATS_METHODDEF
     _SSL__SSLCONTEXT_GET_CA_CERTS_METHODDEF
     _SSL__SSLCONTEXT_GET_CIPHERS_METHODDEF
+    _SSL__SSLCONTEXT_GET_GROUPS_METHODDEF
     _SSL__SSLCONTEXT_SET_PSK_CLIENT_CALLBACK_METHODDEF
     _SSL__SSLCONTEXT_SET_PSK_SERVER_CALLBACK_METHODDEF
     {NULL, NULL}        /* sentinel */
```

**Comment by ronf** [comment]
One additional note: While SSL_CTX_set1_groups_list() has been available since OpenSSL 1.1.1 (the minimum version required by Python right now), some of the other functions didn't come in until later and will probably need to be wrapped in an #ifdef. Specifically, SSL_get0_group_name() was added in OpenSSL 3.2 and SSL_CTX_get0_implemented_groups() is new in OpenSSL 3.5.

Does anyone know how this is typically handled? Does unimplemented functionality like this raise an exception if you try to use it when building Python against an older version of OpenSSL, or does it return some kind of empty/dummy value (depending on the specific return type)? I'm leaning toward an exception here, but I wasn't able to find many examples to model this after.

Another option might be to not define the functions at all when compiling against an older version of OpenSSL, but typically documentation only lists the minimum Python version required for a particular method, not something like the OpenSSL version. So, I don't think that's the right approach.

**Comment by davidkel** [comment]
@ronf I like where you are going with the suggestions and would love to see this get into python (I'm not a maintainer however was looking to contribute but you are ahead of me with this). In the original PR they checked for the version of openssl that comes from the openssl header file, eg
```
#if OPENSSL_VERSION_MAJOR < 3
```
so you should be able to do conditional compilation to either provide a real implementation of the api or return some sort of runtime error saying not supported due to too old openssl for example (or it could just be a no-op at the c layer and a guard implemented in the python ssl wrapper). I think it would require the maintainers to perhaps provide guidance there as to an acceptable approach. 

For further observability would be good to also include the sigalg and peer sigalg used on connection using the `SSL_get_signature_type_nid` and `SSL_get_peer_signature_type_nid`

**Comment by davidkel** [comment]
@ronf I like where you are going with the suggestions and would love to see this get into python (I'm not a maintainer however was looking to contribute but you are ahead of me with this). In the original PR they checked for the version of openssl that comes from the openssl header file, eg
```
#if OPENSSL_VERSION_MAJOR < 3
```
and that seemed to be acceptable at the time, so as to allow python to compile for any version of openssl. As to what to do when it compiles on a version of openssl that can't support the call as you say the question is should it raise an exception for example ? IMO my thinking is that for get calls it could returned an undefined (or a string indicating unavailable) but for the set it should throw an exception. It may also be better to implement the checks in the python ssl wrapper and just have a no-op in the C code. I am not sure the maintainers are following this thread anymore so it may be worth creating a new PR and linking it to the original issue to try to get some visibilty on it. 

For further observability would be good to also include the sigalg and peer sigalg used on connection using the `SSL_get_signature_type_nid` and `SSL_get_peer_signature_type_nid`

**Comment by ronf** [comment]
> @ronf I like where you are going with the suggestions and would love to see this get into python (I'm not a maintainer however was looking to contribute but you are ahead of me with this).

Thanks - I appreciate the support!

As you suggested below, I may try to put together a new PR based on the changes I proposed here, adding in documentation and testing for the new methods. I think the new set_groups() method could potentially even allow us to deprecate the existing set_ecdh_curve(), as you can pass in either the IANA names or the older curve names. It would also cover the ability to specify multiple curves that's covered in this PR.

> In the original PR they checked for the version of openssl that comes from the openssl header file, eg
> 
> ```
> #if OPENSSL_VERSION_MAJOR < 3
> ```
> 
> and that seemed to be acceptable at the time, so as to allow python to compile for any version of openssl.

Yeah - I'd need to use the full OPENSSL_VERSION to be able to match on 3.2 and 3.5, but that's straightforward. My question was only around what the code should do on Python builds linked against older OpenSSL versions.

> As to what to do when it compiles on a version of openssl that can't support the call as you say the question is should it raise an exception for example ? IMO my thinking is that for get calls it could returned an undefined (or a string indicating unavailable) but for the set it should throw an exception. It may also be better to implement the checks in the python ssl wrapper and just have a no-op in the C code. I am not sure the maintainers are following this thread anymore so it may be worth creating a new PR and linking it to the original issue to try to get some visibilty on it.

The set call isn't an issue here, as that has been available since OpenSSL 1.1.1 which is now the minimum supported version for Python. So, it comes down to only the two get calls. I'm thinking an exception might be best here, though, as that would prevent calling code from having to cope with something like an empty string or empty list that normally wouldn't occur. The exception text could also be more specific about what OpenSSL version is needed to get the call to work.

> For further observability would be good to also include the sigalg and peer sigalg used on connection using the `SSL_get_signature_type_nid` and `SSL_get_peer_signature_type_nid`

I agree that could be useful, but since it is somewhat independent of key agreement I'm thinking that might be best handled in a separate PR. Keeping the changes separate could also make the approval process easier.

**Comment by davidkel** [comment]
@ronf do you think the set_ecdh_curve() should call into the set_groups api but still deprecate it so that it can also accept the enhanced openssl format or should it remain as is ? There could be a benefit to existing python code that set the ecdh curve from values in a config file to be able to take advantage by only doing a config change.

**Comment by ronf** [comment]
I'm torn on this. Passing in a single EC curve name to set_groups() should work just fine, but making this change would mean that set_ecdh_curve() accepted multiple curves or groups that were not EC curves, which could break unit tests which might be out there. No existing applications should be trying to pass anything more than a single EC curve name (as anything else wouldn't work today), and if an application/library did want the added functionality it would be really easy to just switch to calling set_groups() in the calling code. While I see what you're saying about taking an opaque string from a config file and being able to change it to specify multiple curves in it without touching the application, I'm not sure that's a common enough use case to be worth the potential risk.

**Comment by ronf** [comment]
Ok - the new issue is #136306, and the new PR is #136307. It now includes updates to the documentation and tests, including marking `set_ecdh_group()` as deprecated. I also settled on raising a NotImplementedError when the version of OpenSSL is too old to support the method being called.

**Comment by ronf** [comment]
#136307 is now merged to main, but the change to deprecate `set_ecdh_group()` has been removed for now. That can be done later as a separate commit.
